src/vivarium_census_prl_synth_pop/components/observers.py

```diff
@@ -1,12 +1,97 @@
-from pathlib import Path
+from abc import ABC, abstractmethod
 
+import pandas as pd
 from vivarium.framework.engine import Builder
 from vivarium.framework.event import Event
+from vivarium.framework.population import PopulationView
 
 from vivarium_census_prl_synth_pop.constants import data_values, metadata
+from vivarium_census_prl_synth_pop.utilities import build_output_dir
 
 
+class BaseObserver(ABC):
+    """Base class for observing and recording relevant state table results. It
+    maintains a separate dataset per concrete observation class and allows for
+    recording/updating on some subset of timesteps (defaults to every time step)
+    and then writing out the results at the end of the sim.
+    """
+
+    def __repr__(self):
+        return "BaseObserver()"
+
+    ##############
+    # Properties #
+    ##############
+
+    @property
+    def name(self):
+        return "base_observer"
+    
+    @property
+    @abstractmethod
+    def output_filename(self):
+        pass
+
+    #################
+    # Setup methods #
+    #################
+
+    def setup(self, builder: Builder):
+        # FIXME: move filepaths to data container
+        # FIXME: settle on output dirs
+        self.output_dir = build_output_dir(builder, subdir="results")
+        self.population_view = self.get_population_view(builder)
+        self.responses = self.get_responses()
+        
+        # Register the listener to update the responses
+        builder.event.register_listener(
+            "collect_metrics",
+            self.on_collect_metrics,
+        )
+        
+        # Register the listener for final write-out
+        builder.event.register_listener(
+        	"simulation_end",
+        	self.on_simulation_end,
+        )
+
+    @abstractmethod
+    def get_population_view(self, builder) -> PopulationView:
+        """Get the population view to be used for observations"""
+        pass
+
+    @abstractmethod
+    def get_responses(self) -> pd.DataFrame:
+        """Initializes the observation/results data structure and schema"""
+        pass
+
+    ########################
+    # Event-driven methods #
+    ########################
+
+    def on_collect_metrics(self, event: Event) -> None:
+        if self.to_observe(event):
+            self.do_observation(event)
+        
+    def to_observe(self, event: Event) -> bool:
+        """If True, will make an observation. This defaults to always True
+        (ie record at every time step) and should be overwritten in each
+        concrete observer as appropriate.
+        """
+        return True
+
+    @abstractmethod
+    def do_observation(self) -> None:
+        """Define the observations in the concrete class"""
+        pass
+
+    def on_simulation_end(self, event: Event) -> None:
+        self.responses.to_hdf(self.output_dir / self.output_filename, key="responses")
```

Comment:
For discussion tomorrow: should the decennial census observer put all of the observations in a single responses dataframe?  In the implementation I am refactoring, it has multiple dataframes for each decennial.